File: JAVA/725. Split Linked List in Parts.java
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        int n = 0;
        ListNode ptr = head;
        while (ptr != null) {
            n++;
            ptr = ptr.next;
        }
        int partSize = n / k;
        int extra = n % k;
        ListNode[] ans = new ListNode[k];        
        ptr = head;
        for (int i = 0; i < k; i++) {
            ans[i] = ptr;
            int currentPartSize = partSize + (extra > 0 ? 1 : 0); 
            extra--;
            for (int j = 0; j < currentPartSize - 1 && ptr != null; j++) {
                ptr = ptr.next;
            }
            if (ptr != null) {
                ListNode nextPart = ptr.next;
                ptr.next = null;
                ptr = nextPart;
            }
        }
        return ans;
    }
}

File: C++/725. Split Linked List in Parts.cpp
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int len=0;
        ListNode*temp=head;
        while(temp!=NULL){
            len++;
            temp=temp->next;
        }
        vector<ListNode*>v(k,NULL);
        int p=len/k;
        int extra=len%k;
        temp=head;
        int j=0;
        while(temp!=NULL){
            ListNode*c=temp;
            ListNode* x=new ListNode(-1);
            ListNode*curr= x;
            for(int i=0;i<p;i++){
                curr->next=new ListNode(temp->val);
                temp=temp->next;
                curr=curr->next;
            }
            if(extra>0){
                curr->next=new ListNode(temp->val);
                temp=temp->next;
                curr=curr->next;
                extra--;
            }
            v[j]=x->next;
            j++;
        }
        return v;
    }
};
